oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/celliterator/CellIterator.java

```diff
@@ -526,74 +545,106 @@ public IterateAllEntry(
-      OSHDBBoundingBox cellBoundingBox = XYGrid.getBoundingBox(new CellId(
-          cell.getLevel(),
-          cell.getId()
-      ), true);
       if (bboxOutsidePolygon.test(cellBoundingBox)) {
         return Stream.empty();
       }
       allFullyInside = bboxInPolygon.test(cellBoundingBox);
     } else {
       allFullyInside = false;
     }
+    return iterateByContribution(cell.getEntities(), allFullyInside);
+  }
+
+  /**
+   * Helper method to easily iterate over all entity modifications in a cell that match a given
+   * condition/filter.
+   *
+   * @param cellData the entities to iterate through
+   * @param allFullyInside
+   *
+   * @return a stream of matching filtered OSMEntities with their clipped Geometries and timestamp
+   *         intervals.
+   */
+  public Stream<IterateAllEntry> iterateByContribution(Iterable<? extends OSHEntity> cellData,
+      boolean allFullyInside) {
+    OSHDBTimestampInterval timeInterval = new OSHDBTimestampInterval(timestamps);
 
     if (includeOldStyleMultipolygons) {
       //todo: remove this by finishing the functionality below
       throw new UnsupportedOperationException("this is not yet properly implemented (probably)");
     }
+    return Streams.stream(cellData)
+        .filter(oshEntity -> allFullyInside || (oshEntity.intersects(boundingBox)
+            && !(isBoundByPolygon && bboxOutsidePolygon.test(oshEntity))))
+        .filter(oshEntityPreFilter)
+        .flatMap(oshEntity -> {
+          var fullyInside = allFullyInside || (
+              oshEntity.coveredBy(boundingBox)
+              && (!isBoundByPolygon || bboxInPolygon.test(oshEntity))
+          );
+          var contribs = new ContributionIterator(oshEntity, timeInterval, fullyInside);
+          return Streams.stream(contribs);
+        });
+  }
 
-    Iterable<? extends OSHEntity> cellData = cell.getEntities();
+  private class ContributionIterator implements Iterator<IterateAllEntry> {
+    private final OSHEntity oshEntity;
+    private final OSHDBTimestampInterval timeInterval;
+    private final boolean fullyInside;
+    private final Map<OSHDBTimestamp, Long> changesetTs;
+    private final List<OSHDBTimestamp> modTs;
+    private final List<OSMEntity> osmEntityAtTimestamps;
 
-    return Streams.stream(cellData).flatMap(oshEntity -> {
-      if (!oshEntityPreFilter.test(oshEntity)
-          || !allFullyInside && (
-              !oshEntity.intersects(boundingBox)
-              || (isBoundByPolygon && bboxOutsidePolygon.test(oshEntity))
-          )) {
-        // this osh entity doesn't match the prefilter or is fully outside the requested
-        // area of interest -> skip it
-        return Stream.empty();
-      }
-      if (Streams.stream(oshEntity.getVersions()).noneMatch(osmEntityFilter)) {
-        // none of this osh entity's versions matches the filter -> skip it
-        return Stream.empty();
```

Comment:
I think I added this in order to speed up queries with filters which cannot exclude whole entities based on the data available in the OSHEntity directly. e.g. `filter("type:relation and type=restriction")` would not be able a OSHRelation which only ever has the tag `type=route` for example. This would then result in unnecessary calls of potentially expensive methods like `getModificationTimestamps` for such OSHEntities.

see 1ad75a71d862ba6d913a0fead7f9550ceccc6386

Reply:
Yes I thought this too but I looked into `getModificationTimestamps` and I was wondering what makes getModificationTimestamps expensive in case when there is no osmEntityFilter match?

[`getModificationTimestamps` ](https://github.com/GIScience/oshdb/blob/439c94b79f10d068001a2a233759e4010698e694/oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java#L175) use the osmEntityFilter to pass it down to
- [`getModificationTimestampsReverseNonRecursed`](https://github.com/GIScience/oshdb/blob/439c94b79f10d068001a2a233759e4010698e694/oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java#L192)
- [`collectMembershipTimeIntervals`](https://github.com/GIScience/oshdb/blob/439c94b79f10d068001a2a233759e4010698e694/oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java#L260)

`getModificationTimestampsReverseNonRecursed` is quite light-weighted as in case of a "noneMatch" it creates just an empty ArrayList.
`collectMembershipTimeIntervals` is also quite light-weighted in case of a "noneMatch".

I agree we could check the result of `getModificationTimestampsReverseNonRecursed` if it is empty [here](https://github.com/GIScience/oshdb/blob/439c94b79f10d068001a2a233759e4010698e694/oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java#L178) to avoid the second filter step of `collectMembershipTimeIntervals`. With this check we have one filter pass in case of a 'noneMatch' and two regular ones otherwise.
If we pre check for a none-match like it was before, we have one filter pass of a "noneMatch" but three in a match case.
Does it make sense, or did I miss something?

Reply:
Ah yes, good point, I forgot about the fact that the filter is passed to `getModificationTimestamps`.

> I agree we could check the result of `getModificationTimestampsReverseNonRecursed` if it is empty here to avoid the second filter step of `collectMembershipTimeIntervals`.

:+1: sounds good!

Reply:
done added check to `getModificationTimestamps`:
```
    var entityTimestamps = getModificationTimestampsReverseNonRecursed(osh, osmEntityFilter);
    if (entityTimestamps.isEmpty()) {
      return entityTimestamps;
    }
```